### ISSUE TITLE
Add support for Eclipse 4.8 test source sets in the classpath file

### ISSUE THREAD

**Issue by danielkaneider** [gradle/gradle#4802]
Eclipse 4.8 M5 added support for test sources (https://www.eclipse.org/eclipse/news/4.8/M5/#JDT). It would be nice if gradle eclipse generation could mark the corresponding entries in the `.classpath` files as such. This could be done among others for source folders, output folders and referenced libraries.

**Comment by howlger** [comment]
The following `build.gradle` snippet works for me as a **workaround** for this issue in Eclipse Photon RC2:

```
apply plugin: 'eclipse'
eclipse.classpath.file.whenMerged {

	// separate output folders required to set the 'test' attribute
	entries.find { it.path == 'src/main/java' }.output = 'bin/main'
	def testSrc = entries.find { it.path == 'src/test/java' }
	testSrc.output = 'bin/test'
	testSrc.entryAttributes['test'] = 'true'

	// libraries visible for test sources only?
	entries.forEach { entry ->
		def entryIn = { it.find { file(entry.path).equals(it) } }
		if (entry.kind == 'lib') {
			entry.entryAttributes['test'] =
				entryIn(configurations.testRuntimeClasspath) &&
				!entryIn(configurations.runtimeClasspath)
		}
	}

}
```

**Comment by eric-milles** [comment]
Current gradle/eclipse plug-in is separating output paths and so I was able to accomplish the same with:
```groovy
eclipse.classpath.file.whenMerged {
  entries.findAll { entry ->
    entry.class.name in [
      'org.gradle.plugins.ide.eclipse.model.Library',
      'org.gradle.plugins.ide.eclipse.model.SourceFolder'
    ]
  }.each { entry ->
    if (entry.entryAttributes['gradle_used_by_scope'] == 'test') {
      entry.entryAttributes['test'] = 'true'
    }
  }
}
```

**Comment by liblit** [comment]
@eric-milles, which *specific* Gradle version, Eclipse version, and Eclipse Buildship plug-in version are you using?

I tried [your suggestion](https://github.com/gradle/gradle/issues/4802#issuecomment-407842647) with Gradle 4.9, Eclipse 4.8.0 (Photon), and Buildship 2.2.1.v20180125-1441. I am working with [a complex Gradle-based project that includes many subprojects](https://github.com/wala/WALA/), so I wrapped your entire Gradle fragment inside an `allprojects { ... }` block to include those many subprojects:

```
allprojects {
  eclipse.classpath.file.whenMerged {
    entries.findAll { entry ->
      entry.class.name in [
        'org.gradle.plugins.ide.eclipse.model.Library',
        'org.gradle.plugins.ide.eclipse.model.SourceFolder'
      ]
    }.each { entry ->
      if (entry.entryAttributes['gradle_used_by_scope'] == 'test') {
        entry.entryAttributes['test'] = 'true'
      }
    }
  }
}
```

Unfortunately, this did not do anything useful. Every entry considered by `entries.findAll` had dynamic type `org.gradle.plugins.ide.eclipse.model.Output`. There were no `Library` or `SourceFolder` entries, which therefore left nothing for the `.each { ... }` stage to operate on, so nothing was marked as being test code.

**Comment by eric-milles** [comment]
@liblit I also have Gradle 4.9, Eclipse 4.8.0, Buildship 2.2.1.  My project is very simple and the `.classpath` looks like this:
```xml
<classpath>
	<classpathentry kind="src" output="bin/main" path="src/main/groovy">
		<attributes>
			<attribute name="gradle_scope" value="main"/>
			<attribute name="gradle_used_by_scope" value="main,test"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="src" output="bin/main" path="src/main/java">
		<attributes>
			<attribute name="gradle_scope" value="main"/>
			<attribute name="gradle_used_by_scope" value="main,test"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="src" output="bin/test" path="src/test/java">
		<attributes>
			<attribute name="gradle_scope" value="test"/>
			<attribute name="gradle_used_by_scope" value="test"/>
			<attribute name="test" value="true"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8/"/>
	<classpathentry kind="con" path="org.eclipse.buildship.core.gradleclasspathcontainer"/>
	<classpathentry kind="output" path="bin/default"/>
</classpath>
```

**Comment by liblit** [comment]
Interesting, @eric-milles. As you point out, we have the same versions of anything that should matter. In the my project’s 28 subproject `.classpath` files I see `classpathentry` elements with a variety of `kind` attribute values: `con` × 56, `output` × 28, and `src` × 38. Yet my `allprojects { eclipse.classpath.file.whenMerged { entries.findAll { ... } } }` code considers 34 entries, all of type `org.gradle.plugins.ide.eclipse.model.Output`.

The fact that there seem to be too many `org.gradle.plugins.ide.eclipse.model.Output` entries is not a big concern. There are indeed 33 subprojects plus the top-level project. Some are not Java projects; apparently these lead to `eclipse.classpath` items but do not ultimately result in `.classpath` files being written. But the complete absence of any `Library` or `SourceFolder` entries is baffling, since clearly `con` and `src` class path elements are making into the `.classpath` files eventually.

**Comment by liblit** [comment]
If I omit `allprojects { ... }` and instead copy @eric-millesʼs [suggested code fragment](https://github.com/gradle/gradle/issues/4802#issuecomment-407842647) directly into each subprojectʼs `build.gradle`, then it works as intended. Good!

But for projects with a large number of subprojects, it would probably not be desirable to clone that fragment into each and every subprojectʼs `build.gradle`. I still have no idea why the `allprojects { ... }` variant only visits `Output` entries and not the others.

**Comment by liblit** [comment]
Here is my version of @eric-millesʼs approach that works with subprojects as well:

``` gradle
allprojects {
  apply plugin: 'eclipse'

  eclipse.classpath.file.whenMerged {
    entries.each {
      if (it in org.gradle.plugins.ide.eclipse.model.AbstractClasspathEntry && it.entryAttributes['gradle_used_by_scope'] == 'test')
        it.entryAttributes['test'] = true
    }
  }
}
```

**Comment by eric-milles** [comment]
I used explicit checks for `Library` and `SourceFolder` because I was unsure if `Container`, `Variable` and `ProjectDependency` should have this applied.

**Comment by donat** [comment]
This has been fixed in https://github.com/gradle/gradle/pull/9484 and available since Gradle 5.6.
